rascal-lsp/src/main/rascal/library/demo/lang/pico/LanguageServer.rsc

```diff
@@ -61,9 +71,46 @@ set[LanguageService] picoLanguageServer(bool allowRecovery) = {
     codeAction(picoCodeActionService),
     rename(picoRenamingService, prepareRenameService = picoRenamePreparingService),
     didRenameFiles(picoFileRenameService),
-    selectionRange(picoSelectionRangeService)
+    selectionRange(picoSelectionRangeService),
+    formatting(picoFormattingService)
 };
 
+list[TextEdit] picoFormattingService(Focus input, FormattingOptions opts) {
+    str original = "<input[-1]>";
+    box = toBox(input[-1]);
+    box = visit (box) { case i:I(_) => i[is=opts.tabSize] }
+    formatted = format(box);
```

Comment:
this code would look better if the options were passed directly to `format` and it would implement them. 

For me the code drops under a certain level of abstraction (under the syntax-directed level) with all the string-level operations like `replaceLast`, `perLine` etc, while format and box2text which indeed take care of these details normally.